src/NuGet.Clients/NuGet.SolutionRestoreManager.Interop/IVsSolutionRestoreService4.cs

```diff
@@ -0,0 +1,28 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace NuGet.SolutionRestoreManager
+{
+    /// <summary>
+    /// Represents a package restore service API for integration with a project system.
+    /// Implemented by NuGet.
+    /// </summary>
+    [ComImport]
+    [Guid("72327117-6552-4685-BD7E-9C40A04B6EE5")]
+    public interface IVsSolutionRestoreService4
+    {
+        /// <summary>
+        /// A project system can call this service (optionally) to register itself to coordinate restore.
+        /// Each project can only register once. NuGet will call into the source to wait for nominations for restore.
+        /// NuGet will remove the registered object when a project is unloaded.
+        /// </summary>
+        /// <param name="restoreInfoSource">Represents a project specific info source</param>
+        /// <param name="cancellationToken">Cancellation token.</param>
+        public Task RegisterRestoreInfoSourceAsync(IVsProjectRestoreInfoSource restoreInfoSource, CancellationToken cancellationToken);
```

Comment:
The XMLDoc is missing information about behaviour, such as what exception is thrown if a project is registered more than once. Also, what happens if a project that does not exist in the solution is provided.

I know there were comments in the linked issue saying that it's non-trivial for a project system to call this API only after IVsSolution knows about the project, but it wasn't clear that it was difficult.  Considering that external customers discover these APIs and start calling it from their own extensions, we have to be mindful that the NuGet team are going to get support requests from customers using 3rd party extensions which are causing NuGet to misbehave, and it will take significant effort for us to discover that the cause was a 3rd party extension. Therefore, I encourage us to start of restrictive, and become more liberal if it turns out to be a problem for our partner, rather than start off too liberal and it causing a support burden.

Reply:
> The XMLDoc is missing information about behaviour, such as what exception is thrown if a project is registered more than once. Also, what happens if a project that does not exist in the solution is provided.

This is a good point. I think we can safely throw an InvalidOperationException if a project has been registered yet. 

At this point, I want to merge this API as we need the project system to implement it so that we're able to test. 
The documentation here is not the end all before this API ships. 
We need to test to confirm that what we're trying to do here actually helps.

Reply:
I added some extra docs based on behavior I am certain of at the time of writing this API. 

This *may* and probably will get tweaked as we progress on the feature.

Reply:
I added some descriptions for the behavior I am confident of as of right now. 

This may need tweaked as we progress in the feature.